okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java

```diff
@@ -754,6 +741,57 @@ private void fixupXOkapiToken(ModuleDescriptor md, MultiMap reqHeaders, MultiMap
     }
   }
 
+  private static void streamHandle(ProxyContext pc, ReadStream<Buffer> readStream,
+                                   WriteStream<Buffer> mainWriteStream,
+                                   List<HttpClientRequest> logWriteStreams) {
+    readStream.handler(data -> {
+      AtomicInteger pend = new AtomicInteger();
+      // two passes.. to avoid drainHandler being fired off too early.
+      // first pass: see if any of writing streams are full?
+      mainWriteStream.write(data);
+      if (mainWriteStream.writeQueueFull()) {
+        pend.incrementAndGet();
+      }
+      for (WriteStream<Buffer> w : logWriteStreams) {
+        w.write(data);
+        if (w.writeQueueFull()) {
+          pend.incrementAndGet();
+        }
+      }
+      if (pend.get() == 0) {
+        return;
+      }
+      // second pass: at least one was full. pause and set up drainHandlers for full ones.
+      readStream.pause();
+      if (mainWriteStream.writeQueueFull()) {
+        mainWriteStream.drainHandler(x -> {
+          if (pend.decrementAndGet() == 0) {
```

Comment:
This won't work in all cases.
A WriteStream will always call the drainHandler, even if ProxyService missed the time period where the WriteStream queue was full.

How about a pauseAndResume method:
```
for each WriteStream
  if not full
    set drainHandler to null
  else
    set drainHandler to pauseAndResume
    pause readStream
    exit this method.
resume readStream
```
Call this method after data has been written to the WriteStreams.